File: LinkedList/GoLang Dsa/Leetcode24.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

type LinkedList struct {
	head *Node
}

func (list *LinkedList) Apend(val int) *Node {

	newNode := &Node{data: val}

	if list.head == nil {

		list.head = newNode
		return list.head
	}

	current := list.head

	for current.next != nil {

		current = current.next
	}

	current.next = newNode

	return list.head
}

func (list *LinkedList) swapPairs(head *Node) *Node {

	if head == nil || head.next == nil {

		return head
	}

	dummyNode := &Node{next: head}
	prev := dummyNode
	slow := list.head
	fast := list.head.next

	for slow != nil && fast != nil {

		slow.next = fast.next
		fast.next = slow
		prev.next = fast

		prev = slow
		slow = slow.next
		if slow != nil {

			fast = slow.next
		}

	}

	return dummyNode.next
}

File: LinkedList/Leetcode24.cpp
#include <iostream>

struct Node {
  int data;
  Node *next;
};

class LinkedList {
  Node *head;

public:
  LinkedList() : head(nullptr) {};

  Node *insertAtEnd(int val) {

    Node *newNode = new Node();
    newNode->data = val;
    newNode->next = nullptr;
    if (!head) {
      head = newNode;
      return head;
    }

    Node *temp = head;
    while (temp) {
      temp = temp->next;
    }
    temp->next = newNode;
    newNode->next = nullptr;

    return head;
  }

  void displayLL() {

    if (!head) {
      std::cout << "List is empty" << std::endl;
    }
    Node *temp = head;
    while (temp) {

      std::cout << temp->data << "->";
      temp = temp->next;
    }
    std::cout << "nullptr" << std::endl;
  }

  Node *getHead() { return head; }
};

class Leetcode24 {};
